### PR TITLE
Testing with pytest

### PR DESCRIPTION
**Pull request by ncclementi**
Add `.md` with testing session following what we showed in the workshop session.
Add pytest entry to `mkdocs.yml`

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -30,6 +30,8 @@ pages:
         - The Python Scientific Stack: jupyter/1.md
     - Laplace:
         - Elliptic PDEs: laplace/1.md
+    - Testing:
+        - Pytest: testing.md
```
Comment (gforsyth) [inline mkdocs.yml]: should be `testing/testing.md`

File: docs/testing/testing.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+`my_dot_product.py`. If you are using `nano` type in a terminal:
+
+```text
+$ nano my_dot_product.py`
+```
+
+Now, we write our dot product function:
+
+```python
+def dot_product (a,b):
```
Comment (gforsyth) [inline docs/testing/testing.md]: no space before parens

File: docs/testing/testing.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    res = 0
+    for i in range(len(a)):
+        res += a[i] * b[i]
+    return res     
+```
+
+Right under our function we can add our first test. What can be a good test?
```
Comment (gforsyth) [inline docs/testing/testing.md]: I think change this to read "What's a good test?"

File: docs/testing/testing.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+Let's try checking if the dot product of zero's vectors returns zero.
```
Comment (gforsyth) [inline docs/testing/testing.md:38]: No `'s` after zero

File: docs/testing/testing.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+```python
+def test_dot_zeros():
+    a=[0,0]
```
Comment (gforsyth) [inline docs/testing/testing.md]: spaces before and after operators

File: docs/testing/testing.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+```
+
+Alright! Now we can start by writing some code and some tests for it. If you
+want to read more about the documentation of `pytest` here is a [link](http://doc.pytest.org/en/latest/contents.html) where you can find all its 
+documentation.
+
+## Testing a dot_product function
+
+First of all let's open a terminal and create a new file called
+`my_dot_product.py`. If you are using `nano` type in a terminal:
```
Comment (gforsyth) [inline docs/testing/testing.md]: below we're running `dot_product.py` but up here it's `my_dot_product.py` -- I'm fine with either but it should be consistent.

File: docs/testing/testing.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+dot_product.py ..
+
+=========================== 2 passed in 0.02 seconds ===========================
+``` 
+
+Do you notice the difference? Now, we collect 2 items (referring to both tests)
+and we have two "." after `dot_product.py` that indicate two passed tests. 
+
+So far our test are passing, but what else can we test, what if we force it to fail? Will our program work if we try to do the dot product of two vectors
+with different length? Let's write a test and try it out.
```
Comment (gforsyth) [inline docs/testing/testing.md:98]: I think this section needs to be reworked. We shouldn't talk about `pytest.raises(AssertionError)` until we've added an assertion to the script. 

I would first add the two tests (`a < b` and `b < a`) like 

```def test_dot_a_lt_b():
    a = [1, 2]
    b = [1, 1, 2]
    assert dot(a, b) == 3
```

the `b < a` test will fail due to an indexing error. 
Then the lesson is that this test has exposed a problem with the function -- we need to add an assertion that the arrays are the same length. Then we can talk about `pytest.raises`

File: docs/testing/testing.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+And they all pass! Why?.
+ 
+Well the test says: I expect an assertion error and it gets an assertion error, so that's a pass!
+
+We can keep writing tests, and improving our function but maybe you have noticed that the first two tests are exactly the same, and last two too. Instead of 
+writing them separately we can use some tools that pytest provide and our tests
+will look cleaner. 
+
+For the first case, where we want the result of the dot product to match an
+expected value we will re-designed our test in the following way:
```
Comment (gforsyth) [inline docs/testing/testing.md]: *redesign

File: docs/testing/testing.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+collected 5 items 
+
+dot_product.py .....
+
+=========================== 5 passed in 0.03 seconds ===========================
+```
+
+## Keeping the tests in a separate file
+
+For a better organization of your code we strongly recomment having your tests
```
Comment (gforsyth) [inline docs/testing/testing.md]: *recommend

File: docs/testing/testing.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+in a separate file. In the example we've been working a way to do that can be:
```
Comment (gforsyth) [inline docs/testing/testing.md:250]: *For the current example we can do the following:

File: docs/testing/testing.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+def dot (a,b):
+    assert len(a)==len(b)  #This is the line we added
+    res = 0
+    for i in range(len(a)):
+        res += a[i] * b[i]
+    return res
+```
+
+In a separate file that can be named `test_dot.py` you put the tests, and the imports that you need, in this
+case you will `import pytest` and `from dot_product import dot`, therefore your file will look likes:
```
Comment (gforsyth) [inline docs/testing/testing.md]: look *like

File: docs/testing/testing.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+    ([1,2],[1,2,3]),
+    ([1,2,3],[42,7]),])
+
+def test_for_assert_error(a,b):
+    with pytest.raises(AssertionError):
+        dot(a,b)
+``` 
+
+Both files should be in the same location and then you just run in you terminal
+`pytest test_dot.py`.
```
Comment (gforsyth) [inline docs/testing/testing.md]: If there's a test file that begins with `test_` then you only need to run `pytest` without any other arguments and pytest will detect it.

**Review by gforsyth — CHANGES_REQUESTED** [review]
Hey @ncclementi -- thanks for putting this together! There are a few typos and inconsistencies that I pointed out and one part that I think we need to rework a little bit.  

Also in the interests of proper Python style, there should be spaces before and after every operator (except `**`) and also spaces after every comma.

File: docs/testing/testing.md (hunk 7 above)
Comment (ncclementi) [reply docs/testing/testing.md:98]: Ok, that sounds better, I was just following what we did in the course but definitely I agree with you. I'll work on it.

**Comment by ncclementi** [comment]
@gforsyth I know this is old, but would you review this so we can merge it? 

Thanks

**Pull request merged**